src/Microsoft.ML.CpuMath/FactorizationMachine/FactorizationMachineInterface.netcoreapp.cs

```diff
@@ -0,0 +1,80 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.Intrinsics.X86;
+using Microsoft.ML.Internal.CpuMath.Core;
+
+namespace Microsoft.ML.Internal.CpuMath
+{
+    [BestFriend]
+    internal static unsafe partial class FieldAwareFactorizationMachineInterface
+    {
+        public static void CalculateIntermediateVariables(int fieldCount, int latentDim, int count, int[] fieldIndices, int[] featureIndices, float[] featureValues,
+            float[] linearWeights, AlignedArray latentWeights, AlignedArray latentSum, ref float response)
+        {
+            Contracts.AssertNonEmpty(fieldIndices);
+            Contracts.AssertNonEmpty(featureValues);
+            Contracts.AssertNonEmpty(featureIndices);
+            Contracts.AssertNonEmpty(linearWeights);
+            Contracts.Assert(Compat(latentWeights));
+            Contracts.Assert(Compat(latentSum));
+
+            unsafe
+            {
+                fixed (int* pf = &fieldIndices[0])
+                fixed (int* pi = &featureIndices[0])
+                fixed (float* px = &featureValues[0])
+                fixed (float* pw = &linearWeights[0])
+                fixed (float* pv = &latentWeights.Items[0])
+                fixed (float* pq = &latentSum.Items[0])
+                fixed (float* pr = &response)
+                {
+                    if (Fma.IsSupported)
+                        FactorizationMachineFmaIntrinsics.CalculateIntermediateVariables(fieldCount, latentDim, count, pf, pi, px, pw, Ptr(latentWeights, pv), Ptr(latentSum, pq), pr);
+                    else if (Avx.IsSupported)
+                        FactorizationMachineAvxIntrinsics.CalculateIntermediateVariables(fieldCount, latentDim, count, pf, pi, px, pw, Ptr(latentWeights, pv), Ptr(latentSum, pq), pr);
+                    else
+                        CalculateIntermediateVariablesNative(fieldCount, latentDim, count, pf, pi, px, pw, Ptr(latentWeights, pv), Ptr(latentSum, pq), pr);
```

Comment:
When Fma and Avx are not supported on the current machine, what do you think about not using native code, but instead falling back to a C# implementation?

The advantage of this is that we can support other architectures (like ARM) without having to compile directly to ARM code.

Reply:
I agree, it would be good to have a fallback implementation to support other architectures. I can open an issue for the same.